### ISSUE TITLE
Redefine RenderStrategyConfig interface and restructure strategies to a more flexible composition

### ISSUE THREAD

**Issue by BioPhoton** [rx-angular/rx-angular#294]
# Problem Solved By The Feature

The current way of providing strategies ends up in an unnecessary amount of possible strategies, unintuitive strategy configuration, and huge configuration options set.
Every feature basically has to end up in the name. E.g. local, localDetach, localChunked, localTimeout

Furthermore we restrict the user from creating custom strategies or at least limit it drastically.
Scheduling would be excluded completely, as well as most of the efficiency techniques.

# Solution

We could stick with strategy names as params for directives and provide ways to register navive as well as custom strategies over e.g. a token.

**New Strategies Names**
The provided default strategy names would be:
- `local`
- `global`
- `noop`
- `native`

**New Strategy Config**
Their configuration objects structure could look like this:

```typescript
interface StrategyCredentials {
  name: string;
  work: (cdRef: ChangedetectorRef, context: any): void,
  behavior: (work: () =>  void, context: any): void 
 }
```

**Default Strategy Credentials**

- `local`
```typescript
const credentials: StrategyCredentials = {
  name: 'local',
  work: (cdRef: ChangeDetectorRef) => cdRef.detectChanges(),
  behavior: (work: () => void) => o$ => o$.pipe(
    coalesceWith(priorityTickMap[SchedulingPriority.animationFrame]),
    tap(() => work())
  );
```

- `global`
```typescript
const credentials: StrategyCredentials = {
  name: 'global',
  work: (cdRef: ChangeDetectorRef) => markDirty(cdRef.context),
  behavior: () => o => o
};
```
- `native`
```typescript
const credentials: StrategyCredentials = {
  name: 'native',
  work: (cdRef) => cdRef.markForCheck(),
  behavior: () => o => o
};
```

- `noop`
```typescript
const credentials: StrategyCredentials = {
  name: 'noop',
  work: () => void 0,
  behavior: () => o => o
};
```

**Custome Strategies:**

Custome strategies can be provided under a token.

```typescript
@NgModule({
  // ...
  providers: [
    {
      provide: RX_RENDER_STRATEGIES,
      useValue: customStrategies,
      multi: true
    }
  ],
  // ...
})
export class AppModule {
}
```

**Custom strategy examples:**

Log strategy
```typescript
const credentials: StrategyCredentials = {
  name: 'log',
  work: (cdRef: ChangeDetectorRef) => {
      cdRef.detectChanges();
      console.log('render'); 
  },
  behavior: (work: () => void) => o$ => o$.pipe(
    tap(() => console.log('update')),
    coalesceWith(priorityTickMap[SchedulingPriority.animationFrame]),
    tap(() => work())
  );
```

Local detach
```typescript
const credentials: StrategyCredentials = {
  name: 'localDetach',
  work: (cdRef: ChangeDetectorRef) => {
      cdRef.reattach();
      cdRef.detectChanges();
      cdRef.detach();
  },
  behavior: (work: () => void) => o$ => o$.pipe(
    coalesceWith(priorityTickMap[SchedulingPriority.animationFrame]),
    tap(() => work())
  );
```

Custom Scheduler Strategy
```typescript
const credentials: StrategyCredentials = {
  name: 'customScheduler',
  work: (cdRef: ChangeDetectorRef) => cdRef.detectChanges(),
  behavior: postTaskBehavior(PostTaskSchedulerPriority.userBlocking)
}
```

**Composition:**
To apply the strategy credentials the composition could look like the following:

```typescript
this.observable$
    .pipe(
        tap((value: any) => this.templateObserver.next(value)),
        rxMaterialize(),
        applyStrategy(this.strategy$, this.cdRef, this.templateManager.getEmbeddedView)
      );

// Very WIP!!
function applyStrategy<T>(
  credentials$: Observable<StrategyCredentials>,
  context,
  getEmbeddedView
): (o$: Observable<RxNotification<T>>) => Observable<RxNotification<T>> {
  return notification$ => notification$.pipe(
    publish((n$) =>
      credentials$.pipe(
        switchMap((credentials) => n$.pipe(
          switchMap(n => {
            const embeddedView = getEmbeddedView(n.kind);
            const work = () => credentials.work(embeddedView, context);
            return of(n).pipe(credentials.behavior(work, embeddedView));
          })
          )
        )
      )
    )
  );
}
```
# Todos
- [x] Introduce `RX_RENDER_STRATEGIES` token
- [x] Helpers to merge `RX_RENDER_STRATEGIES` and default strategies
- [x] Introduce new strategy set into pipes and directives
- [x] Adopt/reconsider `RenderAware` (related to `EmbeddedView`)
- [x] Adopt default strategies
- [ ] Tests
- [ ] API Docs
- [ ] Web Docs
- [x] Organize non default strategies

**Comment by hoebbelsB** [comment]
You point out a good issue here. What I still don't get is, where is the scheduling part? If the base looks like this:

```ts
export function createLocalStrategy<T>(
  config: RenderStrategyFactoryConfig.
): RenderStrategy {

  const behavior = (o) =>
    o.pipe(
       // strategy specific behavior
      config.before$(config.cdRef),
      tap(renderMethod),
      config.after$(config.cdRef)
    );
   // ...
  return {
   // ...
  };
}
```
Where is the scheduling applied? In the `before$` hook?

**Comment by BioPhoton** [comment]
Yes

**Comment by hoebbelsB** [comment]
okay, so the custom logic applied from the user in the `before$` hook is called *after* scheduling finished and *before* renderMethod is actually called, correct?

so very simplified this would look like this?

```ts
schedule();
before();
renderMethod();
after();
```
